### PR TITLE
refactor and support batch pre core for prediction

### PR DESCRIPTION
**Pull request by yangw1234**

### CODE REVIEW

**Comment by yangw1234** [comment]
jenkins pass http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/1020/

File: pyzoo/zoo/pipeline/api/net.py [hunk 1]
```diff
@@ -350,43 +366,71 @@ def __init__(self, loss, optim_method, sess=None):
...
+                 batch_pre_core=-1, hard_code_batch_size=False):
+        if batch_size > 0 and batch_pre_core > 0:
+            raise ValueError("bath_size and batch_per_core should not be set simultaneously")
+
+        node_num, core_num = get_node_and_core_number()
+        self.total_core_num = node_num * core_num
+        if batch_size > 0:
+            if batch_size % self.total_core_num != 0:
                 raise ValueError("batch_size should be a multiple "
                                  "of core_num, but got batch_size: "
```
Comment (jason-dai) [inline pyzoo/zoo/pipeline/api/net.py]: a multiple of total core number

File: pyzoo/zoo/pipeline/api/net.py [hunk 2]
```diff
@@ -350,43 +366,71 @@ def __init__(self, loss, optim_method, sess=None):
...
+        self.batch_pre_core = batch_pre_core
+
+        if not hard_code_batch_size:
+            self.tensors = [tf.placeholder(name=names[i],
+                                           dtype=types[i],
+                                           shape=[None] + shapes[i])
+                            for i in range(len(names))]
         else:
-            batch_pre_core = batch_size / core_num
+            if batch_pre_core is not None:
```
Comment (jason-dai) [inline pyzoo/zoo/pipeline/api/net.py]: batch_pre_core is an Int, so "not None" is always true, yes?

File: pyzoo/zoo/pipeline/api/net.py [hunk 3]
```diff
@@ -445,8 +494,11 @@ def __init__(self, sess, outputs):
         self.inputs = self.dataset.tensors
         _check_the_same(all_required_inputs, self.inputs)
         self.tfnet = TFNet.from_session(sess, self.inputs, outputs)
+        if self.dataset.batch_pre_core <= 0:
+            raise ValueError("You should set batch_pre_core on TFDataset instead of batch_size")
```
Comment (jason-dai) [inline pyzoo/zoo/pipeline/api/net.py]: "for prediction"

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKeras.scala [hunk 4]
```diff
@@ -1287,6 +1302,19 @@ class PythonZooKeras[T: ClassTag](implicit ev: TensorNumeric[T]) extends PythonB
...
 
+  def createTFOptimizer(modelPath: String,
+                        optimMethod: OptimMethod[Float],
+                        x: JavaRDD[Sample],
+                        batchSize: Int = 32): TFOptimizer = {
+    new TFOptimizer(modelPath, optimMethod,
+      toJSample(x).asInstanceOf[RDD[JSample[Float]]], batchSize)
+  }
+
+  def tfOptimize(optimizer: TFOptimizer, endTrigger: Trigger): JList[JTensor] = {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKeras.scala]: is this used?

Comment (yangw1234) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKeras.scala]: No, I forgot to delete it.

File: pyzoo/zoo/pipeline/api/net.py (hunk 2 above)
Comment (yangw1234) [reply pyzoo/zoo/pipeline/api/net.py]: it is a bug, should be
```
if batch_per_core > 0:
```

**Pull request merged**